### DISCUSSION TITLE
Build a SaaS Template

### DISCUSSION THREAD

**Discussion by jacobsfletch** [payloadcms/payload#4395] [Feature Requests & Ideas]
There's been a growing need for a new SaaS starter template, and so it's time we start outlining some of the features and requirements of this work. We came close to shipping a SaaS template earlier this year, but after building out the [Website Template](https://github.com/payloadcms/payload/tree/main/templates/website) and [Ecommerce Template](https://github.com/payloadcms/payload/tree/main/templates/ecommerce), that work was put on hold in order to build out [Payload Cloud](https://payloadcms.com/cloud-pricing) and support one-click deployments. But this timeline wasn't all bad, because we built Payload Cloud _using_ Payload. Needless to say, we learned a lot about subscriptions and recurring payments.

As a starting point, we can duplicate the Website Template, it contains everything we need from page building to front-end rendering (and includes an auth flow) so we can immediately focus on the subscription management aspects of this template. Same goes for the basic Stripe integration from the E-commerce Template (the plugin, product connection, etc.). Much of this can get carried over—but subscriptions are billed entirely differently than one-off payments, so this is where the jumping off point will be for the SaaS template. Fortunately, most of _this_ can be taken directly from our Payload Cloud implementation, so it does not have to be rewritten from scratch. Aaand same goes for the front-end. We can carry over much of the code from [our open-source website](https://github.com/payloadcms/website).

Going off the cuff, need to create some new collections (more details to come):
1. `plans` which are products with recurring payments
1. `teams` which will subscribe to plans using a billing email
1. `users` who will join teams with a given role

Then we need to open up various custom endpoints which our front-end will hit (more details to come):
1. Install the `@payloadcms/plugin-stripe` plugin to make using the Stripe API and handling Stripe webhooks much easier
1. Build a `/create-setup-intent` endpoint for setting up new payment methods on the front-end
1. Build a `/create-subscription` endpoint to verify payment methods and ensure default payments are set, etc

On the `teams` collection, some custom endpoints (more details to come):
1. `/teams/:id/customer` proxy to the related Stripe customer
1. Same with `.../subscriptions`, `.../invoices`, `.../payment-methods`, etc.
1. All the team management related endpoints like `.../accept-invitation`, `.../send-invitation`, etc

As for the front-end, we need to handle:
1. Free trials
1. Checkout
1. Billing management
1. Team and account management

More to come. This is a living document that will be continually updated in greater detail and to better reflect direct feedback from the community. Please join the conversation and/or upvote this discussion if you think this work should be prioritized!

**Comment by octalpixel** [comment]
Love to know how this is going about

**Comment by ydemetriades** [comment]
I would really be interested on this as well. Do we have an update on this? :) 

(I know that you are quite busy right now with v3 and all those goodies that you are preparing!!)

**Comment by grantspilsbury** [comment]
I would love to be able to use a subscription template. Any progress?

**Comment by neudysrd** [comment]
I would really be interested on this as well. Do we have an update on this? 🥹
